### PR TITLE
Add "type: module" (was: Additional Options: UTC Dates + Skip emitting file data)

### PR DESCRIPTION
**Pull request by ikreymer**
This PR adds several options, added via additional options parameters. Not sure if this is best way to add these, or if you would consider adding them, and definitely open to suggestions. Wanted to suggest PR so can avoid having a separate fork. The main purpose of this is to support a use case for creating an ['in-place' ZIP around data that already exists](https://github.com/webrecorder/ipfs-composite-files#in-place-zip) on IPFS.

1. The first option is `utcDates`, which forces using the `Date.getUTC*` functions. Yes, I saw the note re: DOS dates and default local timezone, but it would still be very helpful to have consistent, reproducible date setting, regardless of the time zone. This option allows for that.

2. The next option is `skipEmitFileData`, which may seem a bit strange. The use case is for generating ZIP file for data that is already stored elsewhere (in this case, in IPFS) and emitting only the blocks 'around' the data. In my use case, I use this to detect when the new file entry data starts: https://github.com/webrecorder/ipfs-composite-files/blob/store-ops/src/zip.js#L57 (There's probably more concise way of doing, such as emitting a custom id instead, but wanted to support this with minimal changes to client-zip). Of course, when this is set, the actual zip emitted is no longer valid, as the data is skipped. There's probably a way to support this outside of makeZip, but would require exposing new functions..

3. Also set `type: module` to allow the built index.js to be used with esm modules in node.

Definitely open to thoughts / suggestions for how it might be possible support this use case.

### CODE REVIEW

**Comment by Touffy** [comment]
Hi Ilya.

1. If you want "consistent, reproducible date setting, regardless of the time zone", using UTC in the DOS date is not the way to go. Instead, you can implement the UNIX Extra Field (0x000d) specified in section 4.5.7 of the APPNOTE. Those date fields are regular UNIX timestamps and always interpreted as UTC (unlike the DOS date).

2. I'd like to understand your use case a bit more, but it sounds like client-zip could help if we find a smarter way to output the data. Maybe we could insert some symbols in the async iterable and return it directly instead of its Stream wrapper, so you don't have to rely on pattern matching binary data.

3. Sure.

If you can make a PR with 1 and 3, I could validate that quickly. The `skipEmitFileData` thing maybe needs more discussion.

**Comment by ikreymer** [comment]
> 1. If you want "consistent, reproducible date setting, regardless of the time zone", using UTC in the DOS date is not the way to go. Instead, you can implement the UNIX Extra Field (0x000d) specified in section 4.5.7 of the APPNOTE. Those date fields are regular UNIX timestamps and always interpreted as UTC (unlike the DOS date).

Well, don't actually care about the date value in the ZIP itself, the important part is for ZIP bytes stream to return a consistent stream bytes no matter where it is run, given the same input. But setting `lastModified: new Date(0)` will result in different bytes, based on the timezone, with how the Date class works. I don't see an easy way of doing that without using the `getUTC*` functions. Do you propose a different solution to address that?

> 2. I'd like to understand your use case a bit more, but it sounds like client-zip could help if we find a smarter way to output the data. Maybe we could insert some symbols in the async iterable and return it directly instead of its Stream wrapper, so you don't have to rely on pattern matching binary data.

Actually, yes, I've since had a case where I need the data, but needed to know where it stars, so a zero-length Uint8Array marker at beginning and end of file is a more flexible solution, can then detect these, but the stream is still valid ZIP data.
Can make a separate PR for that later.

**Comment by Touffy** [comment]
I still don't see why you need that "consistent" DOS date in the byte stream. As I said, the correct way to store a date that doesn't depend on system timezone is the UNIX extra field. A DOS date based on UTC will be incorrect on any system that isn't itself set to UTC.

I was actually thinking of using Symbols as markers. You can't distinguish a zero-length buffer from another, but with symbols we can have multiple markers (for start of data, start of headers, etc.) and we insert different symbols in the iteration (and just skip them when converting to a binary stream for normal usage). We could also decorate the initial buffer of a section with a symbol-named property whose value indicates what it's about (and we could even expose structured metadata that way).

**Comment by Touffy** [comment]
If you need the consistency for testing, just set the timezone to UTC when running the test, by prefixing your command line with `TZ=UTC`.

**Comment by ikreymer** [comment]
> If you need the consistency for testing, just set the timezone to UTC when running the test, by prefixing your command line with `TZ=UTC`.

There are many reasons to want to have deterministic output, testing is definitely one of them. It's not always possible to change the system timezone, especially when running in the browser. I am considering using client-zip as part of a library that runs in the browser, where changing the timezone is not an option. (especially since not really using the date at all!)  Would you be opening to an option to just zero out the date field instead, so its always set to a fixed value?

**Comment by ikreymer** [comment]
> There are many reasons to want to have deterministic output, testing is definitely one of them. It's not always possible to change the system timezone, especially when running in the browser. I am considering using client-zip as part of a library that runs in the browser, where changing the timezone is not an option. (especially since not really using the date at all!) Would you be opening to an option to just zero out the date field instead, so its always set to a fixed value?

Ok, I guess I can accomplish deterministic results just adding the offset manually, eg: `lastModified: new Date(time + new Date().getTimezoneOffset() * 60000)`

**Comment by Touffy** [comment]
Yeah, adding the offset will cancel the effect of the system's timezone on the stored date. If you already know a file's modification time, I suppose that means you can predict its DOS timestamp fully. I'd still like to hear what reason you have for that beside testing.
Based on your project's README and what you said, I guess you're running client-zip on the server ? but then you said you'd like to run in the browser, and those timezone issues wouldn't make much sense on the back-end… are you going to access IPFS directly from the browser ?

**Comment by ikreymer** [comment]
> Yeah, adding the offset will cancel the effect of the system's timezone on the stored date. If you already know a file's modification time, I suppose that means you can predict its DOS timestamp fully. I'd still like to hear what reason you have for that beside testing. Based on your project's README and what you said, I guess you're running client-zip on the server ? but then you said you'd like to run in the browser, and those timezone issues wouldn't make much sense on the back-end… are you going to access IPFS directly from the browser ?

Yes, that's just one usage, and eventually the hope is for ipfs-composite-files to also be used in the browser.
Also currently using client-zip in a browser extension.

I've updated the PR to remove utc dates and skipping, instead adding a 'before file start' marker and 'after file end' marker, which can be arbitrary objects, including symbols, that the user passes in. 

Now using it like this:
https://github.com/webrecorder/ipfs-composite-files/blob/store-ops/src/zip.js#L48
and as part of a much more complex system, in the browser:
https://github.com/webrecorder/archiveweb.page/blob/ipfs-custom-splits/src/downloader.js#L259

I'm actually setting the before and after markers to the same object, so can be the same object / symbol for start and end, but made it two markers just in case, for flexibility.. (But could change it to one if you'd prefer)

**Comment by ikreymer** [comment]
> I was actually thinking of using Symbols as markers. You can't distinguish a zero-length buffer from another, but with symbols we can have multiple markers (for start of data, start of headers, etc.) and we insert different symbols in the iteration (and just skip them when converting to a binary stream for normal usage). We could also decorate the initial buffer of a section with a symbol-named property whose value indicates what it's about (and we could even expose structured metadata that way).

I used zero-length Uint8Array as that's backwards compatible with regular streaming (can just be ignored if consuming bytes), but using === can check for the exact marker objects that are passed in. But happy to use a different approach as well!

File: src/zip.ts [hunk 1]
```diff
@@ -36,7 +36,7 @@ export function contentLength(files: Iterable<Metadata>) {
   return centralLength + offset
 }
 
-export async function* loadFiles(files: ForAwaitable<ZipEntryDescription & Metadata>) {
+export async function* loadFiles(files: ForAwaitable<ZipEntryDescription & Metadata>, markerBeforeFileStart: any = null, markerAfterFileEnd: any = null) {
```
Comment (Touffy) [inline src/zip.ts:39]: The second argument should be an Options object, otherwise this will become arguments hell when other options come in.

**Comment by Touffy** [comment]
Here's an example if what I was thinking with symbol decorations, in zip.ts :
```ts
export const zipSectionStart = Symbol()

export async function* fileData(file: ZipFileDescription) {
  // ...
  if (bytes instanceof Uint8Array) {
    bytes[zipSectionStart] = "data"
    yield bytes
  // ...
  } else {
  // ...
    let first = true
    while (true) {
  // ...
      if (first) value[zipSectionStart] = "data"
      first = false
      yield value!
    }    
  }
}
```

You don't need to pass your own zero-length buffers to make that work, or to add empty buffers at all. Just import the symbol and look up the property named with it in each buffer. I didn't add "end" markers because they're a pain to set in a generator loop, but the solution is to always add a marker to the following section (that would be the data descriptor), so when you see a buffer with another marker, you know the previous ("data") section is ended.

**Comment by ikreymer** [comment]
Hm, I guess that could work, but seems like a bit more changes?
But, good news, I realized I could just wrap the generator on my end, and then no changes are needed!

```
async function* addMarkers(iter) {
  yield marker;
  yield *iter;
  yield marker;
}
```

So maybe just 3, `type: module` is all that's needed (in node, works already in the browser).

_Sometimes the best code is no code at all..._

**Comment by ikreymer** [comment]
Closing in favor of cleaner #56 as this is no longer needed.

**Pull request closed without merging**